### PR TITLE
New RuboCop autocorrect

### PR DESCRIPTION
**Pull request by bquorning**
The autocorrection API was changed in Rubocop v0.87.0 (pull request rubocop-hq/rubocop#7868). This PR makes rubocop-rspec compatible with those changes.

The superclass of `RuboCop::Cop::RSpec::Cop` is changed from `::RuboCop::Cop::Cop` to `::RuboCop::Cop::Base`.

This has a few consequences:

- Our `#message` methods get called with a different argument than before. It *can* be customized by defining a `#callback_argument` method, but I think it is clearer to avoid callbacks altogether.
- Our `#autocorrect` methods don't get called anymore. Instead, we extend `Autocorrector`, and the `corrector` is being yielded when calling `#add_offense`, so the code is mostly moved in there. For some cases, this means that some code can be removed, which is nice. For some cases, it means that the methods get too long, or the code complexity gets too high, and in those cases I chose to just call out to an `#autocorrect` method anyway, but of course passing the `corrector` and any usable local variables along.
- Our dependency of RuboCop is bumped from '>= 0.68.1' to '>= 0.87.0'.

See also https://github.com/rubocop-hq/rubocop/blob/master/docs/modules/ROOT/pages/v1_upgrade_notes.adoc for documentation of many of the changes.

We probably need some documentation changes too; I haven’t looked into that yet.

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [ ] Updated documentation.
* [ ] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

File: lib/rubocop/cop/rspec/empty_hook.rb [hunk 1]
```diff
@@ -33,15 +34,11 @@ class EmptyHook < Cop
...
-          end
-        end
-
-        def autocorrect(node)
-          lambda do |corrector|
-            block = node.parent
-            range = range_with_surrounding_space(range: block.loc.expression)
-            corrector.remove(range)
+            add_offense(hook) do |corrector|
+              block = hook.parent
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/empty_hook.rb:38]: block is actually just the `node`, so we can remove this line and use node directly

File: lib/rubocop/cop/rspec/example_wording.rb [hunk 2]
```diff
@@ -53,16 +55,12 @@ def on_block(node)
...
-          lambda do |corrector|
-            corrector.replace(docstring(node), replacement_text(node))
-          end
-        end
-
         private
 
         def add_wording_offense(node, message)
-          add_offense(node, location: docstring(node), message: message)
+          add_offense(docstring(node), message: message) do |corrector|
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/example_wording.rb:61]: maybe save docstring(node) as a variable first and reuse?

File: lib/rubocop/cop/rspec/factory_bot/create_list.rb [hunk 3]
```diff
@@ -51,26 +52,21 @@ def on_block(node)
...
 
           def on_send(node)
             return unless style == :n_times
 
             factory_list_call(node) do |_receiver, _factory, count, _|
               add_offense(
-                node,
-                location: :selector,
+                node.loc.selector,
                 message: format(MSG_N_TIMES, number: count)
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/factory_bot/create_list.rb:66]: lets use a variable for the message

File: lib/rubocop/cop/rspec/context_method.rb [hunk 4]
```diff
@@ -32,13 +34,9 @@ class ContextMethod < Cop
...
           context_method(node) do |context|
-            add_offense(context)
-          end
-        end
-
-        def autocorrect(node)
-          lambda do |corrector|
-            corrector.replace(node.parent.loc.selector, 'describe')
+            add_offense(context) do |corrector|
+              corrector.replace(context.parent.loc.selector, 'describe')
```
[the remark this answers is not in the document]
Comment (bquorning) [reply lib/rubocop/cop/rspec/context_method.rb:38]: Using `node.send_node.loc.selector` instead of `context.parent.loc.selector` works.

File: lib/rubocop/cop/rspec/capybara/feature_methods.rb [hunk 5]
```diff
@@ -72,16 +74,11 @@ def on_block(node)
               next if enabled?(match)
 
               add_offense(
-                send_node,
-                location: :selector,
+                send_node.loc.selector,
                 message: format(MSG, method: match, replacement: MAP[match])
```
[the remark this answers is not in the document]
Comment (bquorning) [reply lib/rubocop/cop/rspec/capybara/feature_methods.rb:78]: `#message` always receives a range now, but I can do this:
```rb
          def message(range)
            match = range.source.to_sym
            format(MSG, method: match, replacement: MAP[match])
          end
```

File: lib/rubocop/cop/rspec/capybara/current_path_expectation.rb [hunk 6]
```diff
@@ -47,30 +49,30 @@ class CurrentPathExpectation < Cop
...
-          def autocorrect(node)
-            lambda do |corrector|
-              return unless node.chained?
+          private
 
-              as_is_matcher(node.parent) do |to_sym, matcher_node|
-                rewrite_expectation(corrector, node, to_sym, matcher_node)
-              end
+          def autocorrect(corrector, node)
+            return unless node.chained?
```
[the remark this answers is not in the document]
Comment (bquorning) [reply lib/rubocop/cop/rspec/capybara/current_path_expectation.rb:61]: This should be `next unless node.chained?` inside the `add_offense` block, right?

File: lib/rubocop/cop/rspec/align_left_let_brace.rb [hunk 7]
```diff
@@ -18,34 +18,31 @@ module RSpec
...
+            end
           end
         end
 
         private
 
         def token_aligner
-          @token_aligner ||=
-            RuboCop::RSpec::AlignLetBrace.new(processed_source.ast, :begin)
+          RuboCop::RSpec::AlignLetBrace.new(processed_source.ast, :begin)
```
[the remark this answers is not in the document]
Comment (bquorning) [reply lib/rubocop/cop/rspec/align_left_let_brace.rb:45]: Yeah, just removing the memoization probably isn’t the best approach. How about I just move the instantiation into the `#on_new_investigation` method?

**Comment by bquorning** [comment]
CodeClimate complains about some added complexity, which should probably be addressed. But I think it’s important that we get a new release out pretty quickly, because right now we are incompatible with the most recent version of RuboCop.

**Comment by marcandre** [comment]
There's a lot of activity here 👍
Let me know if I should review this now, or ping me when the dust settles 😄

File: lib/rubocop/cop/rspec/factory_bot/create_list.rb [hunk 8]
```diff
@@ -51,26 +52,19 @@ def on_block(node)
...
-            add_offense(node.send_node, message: MSG_CREATE_LIST)
+            add_offense(node.send_node, message: MSG_CREATE_LIST) do |corrector|
+              CreateListCorrector.new(node.send_node).call(corrector)
+            end
           end
 
           def on_send(node)
             return unless style == :n_times
 
             factory_list_call(node) do |_receiver, _factory, count, _|
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/factory_bot/create_list.rb:63]: it seems that we never use the other captures, so we can just remove them from the pattern

File: lib/rubocop/cop/rspec/hook_argument.rb [hunk 9]
```diff
@@ -100,10 +94,12 @@ def check_implicit(method_send)
           return if implicit_style?
 
           add_offense(
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/hook_argument.rb:96]: Can you make this call one line as well?

File: lib/rubocop/cop/rspec/implicit_expect.rb [hunk 10]
```diff
@@ -55,19 +56,12 @@ def on_send(node) # rubocop:disable Metrics/MethodLength
             opposite_style_detected
 
             add_offense(
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/implicit_expect.rb:58]: one line if possible :pray:

File: lib/rubocop/cop/rspec/leading_subject.rb [hunk 11]
```diff
@@ -43,27 +45,25 @@ def on_block(node)
...
-            RuboCop::RSpec::Corrector::MoveNode.new(
-              node, corrector, processed_source
-            ).move_before(first_node)
-          end
-        end
-
         private
 
+        def autocorrect(corrector, node)
+          first_node = find_first_offending_node(node)
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/leading_subject.rb:61]: In a future refactoring, we can save the offending node in the check_previous_node method and just pass it to the autocorrect

File: lib/rubocop/cop/rspec/predicate_matcher.rb [hunk 12]
```diff
@@ -301,6 +270,7 @@ def replacement_matcher(node)
       #   # good - the above code is rewritten to it by this cop
       #   expect(foo.something?).to be_truthy
       class PredicateMatcher < Cop
+        extend AutoCorrector
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/predicate_matcher.rb:273]: Why is this needed?

File: lib/rubocop/cop/rspec/return_from_stub.rb [hunk 13]
```diff
@@ -59,24 +60,17 @@ def on_block(node)
...
-            BlockBodyCorrector.new(node)
-          end
-        end
-
         private
 
         def check_and_return_call(node)
           and_return_value(node) do |and_return, args|
             unless dynamic?(args)
               add_offense(
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/return_from_stub.rb:68]: one line :pray:

File: lib/rubocop/cop/rspec/return_from_stub.rb [hunk 14]
```diff
@@ -85,10 +79,11 @@ def check_block_body(block)
           body = block.body
           unless dynamic?(body) # rubocop:disable Style/GuardClause
             add_offense(
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/return_from_stub.rb:81]: one line :pray:

File: lib/rubocop/cop/rspec/it_behaves_like.rb [hunk 15]
```diff
@@ -28,14 +29,12 @@ class ItBehavesLike < Cop
 
         def on_send(node)
           example_inclusion_offense(node, alternative_style) do
-            add_offense(node)
+            add_offense(node) do |corrector|
```
Comment (mobilutz) [inline lib/rubocop/cop/rspec/it_behaves_like.rb:32]: Shouldn't there be a `message` parameter to `add_offense`?
```suggestion
            add_offense(node, message: format(MSG, replacement: style, original: alternative_style)) do |corrector|
```

File: lib/rubocop/cop/rspec/not_to_not.rb [hunk 16]
```diff
@@ -24,14 +25,12 @@ class NotToNot < Cop
 
         def on_send(node)
           not_to_not_offense(node, alternative_style) do
-            add_offense(node, location: :selector)
+            add_offense(node.loc.selector) do |corrector|
```
Comment (mobilutz) [inline lib/rubocop/cop/rspec/not_to_not.rb:28]: The `add_offense` should have the `message` parameter:
```suggestion
            add_offense(node.loc.selector, message: format(MSG, replacement: style, original: alternative_style)) do |corrector|
```

**Comment by bquorning** [comment]
@Darhazer I won’t have time to work on this today. If you have time could you please squash, merge and release? If not, I’ll look at it tomorrow.

File: lib/rubocop/cop/rspec/not_to_not.rb (hunk 16 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/not_to_not.rb:28]: If no `message` parameter is passed, RuboCop will call the `#message` method which returns a string. I prefer the more explicit version, but the callback approach still works.

Comment (mobilutz) [reply lib/rubocop/cop/rspec/not_to_not.rb:28]: Ok, got it.
I was just checking the rest of the Changelog, and almost everywhere the code is now calling `add_offense` with the `message` parameter, except when there is no interpolation, and the `MSG` constant is being used automatically.

Maybe this is something, that can be aligned afterwards as well.
So that the code is in line with the expectation from [`add_offense`](https://github.com/rubocop-hq/rubocop/blob/v0.87.1/lib/rubocop/cop/internal_affairs/redundant_message_argument.rb#L12-L21)

File: lib/rubocop/cop/rspec/factory_bot/create_list.rb (hunk 8 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/factory_bot/create_list.rb:63]: I’ll fix that in a separate PR.

File: lib/rubocop/cop/rspec/predicate_matcher.rb (hunk 12 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/predicate_matcher.rb:273]: A cop must `extend AutoCorrector` to be able to autocorrect: https://github.com/rubocop-hq/rubocop/blob/v0.87.1/lib/rubocop/cop/base.rb#L304

**Comment by marcandre** [comment]
Wow, impressive PR 💪!

**Comment by bquorning** [comment]
Thanks. I hope I got everything right 😅

File: lib/rubocop/cop/rspec/predicate_matcher.rb (hunk 12 above)
Comment (Darhazer) [reply lib/rubocop/cop/rspec/predicate_matcher.rb:273]: Oh, I didn't notice this is the actual cop, as it is like 270 lines below the start of the code, though it was a helper class

Comment (Darhazer) [reply lib/rubocop/cop/rspec/predicate_matcher.rb:273]: Still somehow confusing, as the offsense/correction is provided by the included modules. Do we need the `autocorrect` method in this class?

Comment (marcandre) [reply lib/rubocop/cop/rspec/predicate_matcher.rb:273]: Shouldn't it be `PredicateMatcher < Base`?

If is it reused, the `module` should probably do it itself:
```
def self.included(base)
  base.extend AutoCorrector
end
```

File: lib/rubocop/cop/rspec/factory_bot/create_list.rb (hunk 8 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/factory_bot/create_list.rb:63]: See #959.

**Comment by pirj** [comment]
Sleek! 👏 
Thanks a lot guys.

File: lib/rubocop/cop/rspec/leading_subject.rb (hunk 11 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/leading_subject.rb:61]: #960
